spark-jobs/src/main/scala/za/co/absa/enceladus/standardization/StandardizationJob.scala

```diff
@@ -424,53 +425,58 @@ object StandardizationJob {
   }
 
   /**
-    * Adds metadata about the number of records in raw data by checking Atum's checkpoints first.
-    * If raw record count is not available in checkpoints the method will calculate that count
-    * based on the provided raw dataframe.
+    * Adds metadata about the number of records in raw and source data by checking Atum's checkpoints first.
     *
     * @return The number of records in a checkpoint corresponding to raw data (if available)
     */
-  private def addRawRecordCountToMetadata(df: DataFrame): Unit = {
+  private def addRawAndSourceRecordCountsToMetadata: Unit = {
     val checkpointRawRecordCount = getRawRecordCountFromCheckpoints
+    val checkpointSourceRecordCount = getSourceRecordCountFromCheckpoints
 
-    val rawRecordCount = checkpointRawRecordCount match {
-      case Some(num) => num
-      case None      => df.count
+    (checkpointRawRecordCount, checkpointSourceRecordCount) match {
+      case (Right(x), Right(y)) => {
+        Atum.setAdditionalInfo(s"raw_record_count" -> x.toString)
+        Atum.setAdditionalInfo(s"source_record_count" -> y.toString)
+      }
+      case (Right(_), Left(er)) => throw new ValidationException("Field errors", Seq(er))
+      case (Left(er), Right(_)) => throw new ValidationException("Field errors", Seq(er))
+      case (Left(er1), Left(er2)) => throw new ValidationException("Field errors", Seq(er1,er2))
     }
-    Atum.setAdditionalInfo(s"raw_record_count" -> rawRecordCount.toString)
   }
 
   /**
     * Gets the number of records in raw data by traversing Atum's checkpoints.
     *
     * @return The number of records in a checkpoint corresponding to raw data (if available)
     */
-  private def getRawRecordCountFromCheckpoints: Option[Long] = {
+  private def getRawRecordCountFromCheckpoints: Either[String, Long] = {
+    getFieldRawRecordFromCheckpoints("row")
+  }
+
+  /**
+   * Gets the number of records in source data by traversing Atum's checkpoints.
+   *
+   * @return The number of records in a checkpoint corresponding to source data (if available)
+   */
+  private def getSourceRecordCountFromCheckpoints: Either[String, Long] = {
+    getFieldRawRecordFromCheckpoints("source")
+  }
+
+  private def getFieldRawRecordFromCheckpoints(field: String) = {
```

Comment:
This method can encounter 3 types of error conditions. But It seems it distinguishes only 2 of them.
The 3 error conditions are:
- No checkpoint that has the name `name`
- There is a checkpoint `name`, but it doesn't have a `count` measurement
- There is a checkpoint `name`, it has a `count` measurement, but `count<0`

Reply:
For this kind of use case, the Validated class would be quite a good fit